src/server/service/kv.rs

```diff
@@ -211,12 +215,26 @@ impl<E: Engine, L: LockManager, F: KvFormat> Service<E, L, F> {
     }
 }
 
+macro_rules! reject_if_cluster_id_mismatch {
+    ($req:expr, $self:ident, $ctx:expr, $sink:expr) => {
+        let req_cluster_id = $req.get_context().get_cluster_id();
+        if req_cluster_id > 0 && req_cluster_id != $self.cluster_id {
+            // Reject the request if the cluster IDs do not match.
+            warn!("unexpected request with different cluster id is received"; "req" => ?&$req);
+            let e = RpcStatus::new(RpcStatusCode::INVALID_ARGUMENT);
+            $ctx.spawn($sink.fail(e).unwrap_or_else(|_| {}),);
```

Comment:
Should we use gRPC's error instead of returning our own error? Is it enough for the client to know the reason of failing? I'm afraid the client may treat it as other RPC errors and then continue sending unrelated stuff like health checks and retrying requests

Reply:
We may check the status code [here](https://github.com/tikv/client-go/blob/bfab41501bc57bf7b575b070da8796e1edbf9e63/internal/locate/region_request.go#L1869) later?

Reply:
@zyguan Should we assume that the `INVALID_ARGUMENT` error code always stands for cluster id mismatching? I'm afraid it's not very correct in semantics.

Reply:
According to the [code definition](https://grpc.github.io/grpc/core/md_doc_statuscodes.html), `INVALID_ARGUMENT` LGTM. We can also attach [a message](https://github.com/grpc/grpc/blob/master/src/proto/grpc/status/status.proto#L87) to show the detailed reason?